akka-actor/src/main/scala/akka/actor/dungeon/DeathWatch.scala

```diff
@@ -63,7 +64,19 @@ private[akka] trait DeathWatch { this: ActorCell =>
   protected def receivedTerminated(t: Terminated): Unit =
     terminatedQueued.get(t.actor).foreach { optionalMessage =>
       terminatedQueued -= t.actor // here we know that it is the SAME ref which was put in
-      receiveMessage(optionalMessage.getOrElse(t))
+      optionalMessage match {
+        case Some(custom) =>
+          // needed for stashing of custom watch messages to work (or stash will stash the Terminated message instead)
+          val originalMessage = currentMessage
+          currentMessage = Envelope(custom, sender())
```

Comment:
This is the fix

Reply:
we'll have to check that this doesn't break Cinnamon context propagation (might be something attached to the `Envelope`)

Reply:
👍

Reply:
If you use copy instead of a new instance then it'll work: https://github.com/lightbend/cinnamon/blob/master/cinnamon-agent-and-friends/cinnamon-akka-spi-implementation/src/main/scala/akka/dispatch/Envelope.scala#L43